CRM/MembershipExtras/Hook/Alter/CalculatedMembershipStatus.php

```diff
@@ -0,0 +1,336 @@
+<?php
+
+/**
+ * Calculates membership status taking into account in_arrears and not_arrears
+ * start and end events.
+ */
+class CRM_MembershipExtras_Hook_Alter_CalculatedMembershipStatus {
+
+  private $membership;
+  private $calculationArguments;
+  private static $memberShipStatuses = array();
+  private static $contributionStatusValueMap = array();
+
+  public function __construct() {
+    if (count(self::$memberShipStatuses) == 0) {
+      $membershipStatuses = civicrm_api3('MembershipStatus', 'get', [
+        'sequential' => 1,
+        'options' => ['sort' => 'weight ASC', 'limit' => 0],
+      ]);
+      self::$memberShipStatuses = $membershipStatuses['values'];
+    }
+
+    if (count(self::$contributionStatusValueMap) == 0) {
+      $contributionStatuses = civicrm_api3('OptionValue', 'get', array(
+        'option_group_id' => "contribution_status",
+      ));
+
+      foreach ($contributionStatuses['values'] as $currentStatus) {
+        self::$contributionStatusValueMap[$currentStatus['name']] = $currentStatus['value'];
+      }
+    }
+  }
+
+  /**
+   * This hook is called when membership status is being calculated.
+   *
+   * @param array $calculatedStatus
+   *   Membership status details as determined by CiviCRM core
+   * @param array $arguments
+   *   Arguments passed in to calculate date
+   * @param array $membership
+   *   Membership details from the calling function
+   */
+  public function alterMembershipStatus(&$calculatedStatus, $arguments, $membership) {
+    $this->membership = $membership;
+    $this->calculationArguments = $arguments;
+
+    foreach (self::$memberShipStatuses as $currentStatus) {
+      $startEventIsArrearsRelated = stripos($currentStatus['start_event'], 'arrears') !== false;
+      $endEventIsArrearsRelated = stripos($currentStatus['end_event'], 'arrears') !== false;
+
+      if (!$startEventIsArrearsRelated && !$endEventIsArrearsRelated) {
+        // If we reach calculated status, we don't need to consider other options by weight.
+        if ($calculatedStatus['id'] == $currentStatus['id']) {

```

Comment:
Statuses are verified in order by their weight, and once a status that meets start and end criteria is found, the rest are ignored. The only reason to check a status in this hook is if it uses arrears related events for either start or end, in case it was falsely determined it did not meet start and end criteria.

Since in this case: 
- I have already processed all past statuses and found that none of them meet start and end events
- The current status does not use 'In arrears' or 'Not in arrears' events
- CiviCRM has already evaluated that this status meets both start and end criteria

I can safely deduce this is the actual status of the membership and there is no need to continue processing the rest of possible statuses, as they should only be evaluated if all past statuses do not meet start and end criteria.